File: src/generation/outputFormatter.ts
import ExcelJS from "exceljs";
import { GeneratedTestCase, QualityReport } from "../core/types";

/**
 * Renders generated test cases into the supported delivery formats.
 */
export class OutputFormatter {
  formatPlaywright(testCases: GeneratedTestCase[]): string {
    const header = `import { test, expect } from '@playwright/test';\n\n`;
    const body = testCases
      .map((tc) => {
        const comment = `// ${tc.skillId} - ${tc.metadata.testName}`;
        return `${comment}\n${this.extractTestCode(tc.code)}`;
      })
      .join("\n\n");
    return header + body + "\n";
  }

  formatJSON(testCases: GeneratedTestCase[]): string {
    const payload = {
      metadata: {
        generatedAt: new Date().toISOString(),
        totalTests: testCases.length,
        framework: testCases[0]?.framework ?? "playwright",
      },
      testCases: testCases.map((tc) => ({
        skillId: tc.skillId,
        testName: tc.metadata.testName,
        steps: tc.metadata.steps,
        assertions: tc.metadata.assertions,
        expectedResult: tc.metadata.expectedResult,
        tags: tc.metadata.tags,
        priority: tc.metadata.priority,
        score: tc.evaluation?.overallScore,
        code: tc.code,
      })),
    };
    return JSON.stringify(payload, null, 2);
  }

  async formatXLSX(testCases: GeneratedTestCase[]): Promise<Buffer> {
    const wb = new ExcelJS.Workbook();
    wb.creator = "test-case-generator";
    const sheet = wb.addWorksheet("Test Cases");

    sheet.columns = [
      { header: "Skill ID", key: "skillId", width: 18 },
      { header: "Test Name", key: "testName", width: 48 },
      { header: "Steps", key: "steps", width: 60 },
      { header: "Expected Result", key: "expected", width: 40 },
      { header: "Priority", key: "priority", width: 12 },
      { header: "Tags", key: "tags", width: 24 },
      { header: "Score", key: "score", width: 10 },
    ];
    sheet.getRow(1).font = { bold: true };

    for (const tc of testCases) {
      sheet.addRow({
        skillId: tc.skillId,
        testName: tc.metadata.testName,
        steps: tc.metadata.steps
          .map((s, i) => `${i + 1}. ${s}`)
          .join("\n"),
        expected: tc.metadata.expectedResult ?? "",
        priority: tc.metadata.priority,
        tags: tc.metadata.tags.join(", "),
        score: tc.evaluation?.overallScore ?? "",
      });
    }
    sheet.eachRow((row) => (row.alignment = { wrapText: true, vertical: "top" }));

    const arrayBuffer = await wb.xlsx.writeBuffer();
    return Buffer.from(arrayBuffer);
  }

  formatReport(report: QualityReport): string {
    return JSON.stringify(report, null, 2);
  }

  private extractTestCode(code: string): string {
    const cleaned = code.replace(/```(?:[a-z]+)?/gi, "").trim();
    return cleaned;
  }
}

File: java/src/main/java/io/github/ankitat/qaforge/OutputFormatter.java
package io.github.ankitat.qaforge;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Renders generated test cases into Playwright source or JSON metadata. */
public final class OutputFormatter {

  public String formatPlaywright(List<GeneratedTestCase> testCases) {
    StringBuilder sb = new StringBuilder("import { test, expect } from '@playwright/test';\n\n");
    for (GeneratedTestCase tc : testCases) {
      sb.append("// ").append(tc.skillId).append(" - ").append(tc.metadata.testName()).append('\n');
      sb.append(stripFences(tc.code)).append("\n\n");
    }
    return sb.toString();
  }

  public String formatJson(List<GeneratedTestCase> testCases) {
    List<Map<String, Object>> cases = new ArrayList<>();
    for (GeneratedTestCase tc : testCases) {
      Map<String, Object> m = new LinkedHashMap<>();
      m.put("skillId", tc.skillId);
      m.put("testName", tc.metadata.testName());
      m.put("steps", tc.metadata.steps());
      m.put("assertions", tc.metadata.assertions());
      m.put("tags", tc.metadata.tags());
      m.put("priority", tc.metadata.priority());
      m.put("score", tc.evaluation != null ? tc.evaluation.overallScore() : null);
      m.put("code", tc.code);
      cases.add(m);
    }
    Map<String, Object> payload = new LinkedHashMap<>();
    Map<String, Object> meta = new LinkedHashMap<>();
    meta.put("totalTests", testCases.size());
    meta.put("framework", "playwright");
    payload.put("metadata", meta);
    payload.put("testCases", cases);
    return Json.stringify(payload);
  }

  private static String stripFences(String code) {
    return code.replaceAll("```(?:[a-zA-Z]+)?", "").trim();
  }
}
